src/Functions/overlay.cpp

```diff
@@ -0,0 +1,735 @@
+
+            if constexpr (!is_utf8)
+            {
+                size_t new_res_size = res_data.size() + prefix_size + replace_size + suffix_size + 1; /// +1 for zero terminator
+                res_data.resize(new_res_size);
+
+                /// copy prefix before replaced region
+                memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], &input_data[input_offset], prefix_size);
+                res_offset += prefix_size;
+
+                /// copy replace
+                memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], &replace_data[replace_offset], replace_size);
+                res_offset += replace_size;
+
+                /// copy suffix after replaced region. It is not necessary to copy if suffix_size is zero.
+                if (suffix_size)
+                {
+                    memcpySmallAllowReadWriteOverflow15(
+                        &res_data[res_offset], &input_data[input_offset + prefix_size + valid_length], suffix_size);
+                    res_offset += suffix_size;
+                }
+            }
+            else
+            {
+                const auto * input_begin = &input_data[input_offset];
+                const auto * input_end = &input_data[input_offset + input_bytes];
+                const auto * prefix_end = UTF8StringSource::skipCodePointsForward(input_begin, prefix_size, input_end);
+                size_t prefix_bytes = prefix_end > input_end ? input_bytes : prefix_end - input_begin;
+                const auto * suffix_begin = UTF8StringSource::skipCodePointsBackward(input_end, suffix_size, input_begin);
+                size_t suffix_bytes = input_end - suffix_begin;
+                size_t new_res_size = res_data.size() + prefix_bytes + replace_bytes + suffix_bytes + 1; /// +1 for zero terminator
+                res_data.resize(new_res_size);
+
+                /// copy prefix before replaced region
+                memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], input_begin, prefix_bytes);
+                res_offset += prefix_bytes;
+
+                /// copy replace
+                memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], &replace_data[replace_offset], replace_bytes);
+                res_offset += replace_bytes;
+
+                /// copy suffix after replaced region. It is not necessary to copy if suffix_bytes is zero.
+                if (suffix_bytes)
+                {
+                    memcpySmallAllowReadWriteOverflow15(&res_data[res_offset], suffix_begin, suffix_bytes);
+                    res_offset += suffix_bytes;
+                }
+            }
+
+            /// add zero terminator
+            res_data[res_offset] = 0;
+            ++res_offset;
+            res_offsets[i] = res_offset;
+        }
+    }
+};
+
+}
+
+REGISTER_FUNCTION(Overlay)
+{
+    factory.registerFunction<FunctionOverlay<false>>({}, FunctionFactory::Case::Insensitive);
```

Comment:
Needs in-source function documentation, for an example see the end of src/Functions/makeDate.cpp.